std/src/main/scala/algebra/std/list.scala

```diff
@@ -11,15 +12,21 @@ trait ListInstances {
 }
 
 class ListOrder[A](implicit ev: Order[A]) extends Order[List[A]] {
-  def compare(x: List[A], y: List[A]): Int =
-    (x, y) match {
-      case (Nil, Nil) => 0
-      case (Nil, _) => -1
-      case (_, Nil) => 1
-      case (x :: xs, y :: ys) =>
-        val n = ev.compare(x, y)
-        if (n != 0) n else compare(xs, ys)
-    }
+  def compare(xs: List[A], ys: List[A]): Int = {
+    @tailrec def loop(xs: List[A], ys: List[A]): Int =
+      xs match {
+        case Nil =>
```

Comment:
FWIW, I benchmarked similar code a couple years ago at Precog and generally found that building a tuple and immediately matching on it was generally as fast as doing 2 matches. The code generated by Scala, at the time, was what you'd expect (allocate a Tuple2, then destructure it). Not sure if the JVM was inlining it or if stack allocation optimization kicked in. But, without benchmarks, I wouldn't want to commit to anything else here :\